File: python/packages/middlewares/src/package_middlewares/tenantDb.py
import re
from fastapi import HTTPException, Request, status
from typing import Optional

from .types import JwtClaims


async def tenant_db(request: Request):
    claims: Optional[JwtClaims] = request.state.claims

    iss = claims.iss if claims else None
    realm = None
    if iss:
        match = re.search(r"/realms/([^/]+)$", iss)
        realm = match.group(1) if match else None

    tenant = request.headers.get("x-tenant")

    if not realm or not tenant:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="Missing tenant info")

    request.state.realm = realm
    request.state.tenant = tenant

File: node/packages/middlewares/src/tenantDb.ts
import { Request, Response, NextFunction } from "express";

/**
 * Middleware: Attach tenant-specific DB instance to the request
 */
export function tenantDb(req: Request, res: Response, next: NextFunction) {
  // realm comes from verified token claims
  const iss: string | undefined = (req.claims as any)?.iss;
  const match = iss?.match(/\/realms\/([^/]+)$/);
  const realm = match ? match[1] : null;

  // tenant is expected as a header
  const tenant = req.headers["x-tenant"] as string;

  if (!realm || !tenant) {
    return res.status(400).send("Missing tenant info");
  }

  req.realm = realm;
  req.tenant = tenant;
  next();
}
